packages/offix-datastore/src/DataStore.ts

```diff
@@ -1,76 +1,84 @@
+ * and GraphQL replication engine.
+ */
 export class DataStore {
-  public networkStatus: any;
-  private dbName: string;
-  private schemaVersion: number;
   private models: Model<unknown>[];
-  private indexedDB: IndexedDBStorageAdapter;
   private storage: LocalStorage;
-  private clientConfig: any;
-  private metadataName = "metadata";
 
-  constructor(config: DataStoreConfig) {
-    this.dbName = config.dbName;
-    this.schemaVersion = config.schemaVersion || 1; // return 1 if schemaVersion is undefined or 0
-    this.clientConfig = config.clientConfig;
+  private config: DataStoreConfig;
+  private engines: CustomEngines | undefined;
+
+  constructor(config: DataStoreConfig, engines?: CustomEngines) {
+    this.config = config;
     this.models = [];
-    this.indexedDB = new IndexedDBStorageAdapter();
-    this.storage = new LocalStorage(this.indexedDB);
-    this.indexedDB.addStore({ name: this.metadataName });
+    this.engines = engines;
+
+    if (engines && engines.storeAdapter) {
+      this.storage = new LocalStorage(engines.storeAdapter);
+    } else {
+      const indexedDB = new IndexedDBStorageAdapter();
+      this.storage = new LocalStorage(indexedDB);
+    }
   }
 
   public createModel<T>(schema: DataSyncJsonSchema<T>) {
     const modelSchema = new ModelSchema(schema);
-    const model = new Model<T>(modelSchema, this.storage, this.metadataName);
-    this.models.push(model);
+    const model = new Model<T>(modelSchema, this.storage);
+    // TODO replace any
+    this.models.push(model as any);
     return model;
   }
 
   public init() {
-    // TODO rename createGraphQLClient since it is responsible
-    // for the creation of the network status handler too
-    const { gqlClient, networkStatus } = createGraphQLClient(this.clientConfig);
-    this.networkStatus = networkStatus;
-    const queries = buildGraphQLCRUDQueries(this.models);
-    const gqlReplicator = new GraphQLCRUDReplicator(gqlClient, queries);
-    const engine = new MutationReplicationEngine(gqlReplicator, (this.storage as LocalStorage));
-    this.pushReplicator(gqlReplicator);
-    this.indexedDB.createStores(this.dbName, this.schemaVersion);
-    engine.start();
-  }
+    // Created fixed stores for replication
+    // TODO this is just workaround for replication engine
+    this.storage.addStore({ name: MUTATION_QUEUE, keyPath: MUTATION_QUEUE_KEY });
+    this.storage.addStore({ name: MODEL_METADATA, keyPath: MODEL_METADATA_KEY });
+    // TODO this fails on firefox
+    this.storage.createStores(this.config);
 
-  private pushReplicator(replicator: IReplicator) {
-    this.models.forEach((model) => {
-      model.setReplicator(replicator);
-    });
+    if (this.config.replicationConfig) {
+      if (this.engines?.replicator) {
+        this.engines?.replicator.start(this.models, this.storage);
+      } else {
+        const gqlReplicator = new GraphQLCRUDReplicator(this.config.replicationConfig);
```

Comment:
If we are planning to expose the replication engine, then maybe we can set the gqlReplicator as the default in the constructor if no replication engine is provided?